### PR TITLE
[User Model] [Fix] Update existing subscription on token changes instead of creating new

### PR DESCRIPTION
**Pull request by jkasten2**
# Description
## 1 Line Summary
Fix bug where new push subscriptions were being created instead of updating the existing one.

## Details
There are two key changes that were needed to fix this bug:
1. `getCurrentPushSubscriptionModel()`  - wasn't correctly using `lastKnownPushToken`, fixed in commit e9134eec3fb7c9f12f7520b6718e95740738bf53
2. `checkAndTriggerSubscriptionChanged()` - wasn't checking for token changes, so the backend was never updated, fixed in commit https://github.com/OneSignal/OneSignal-Website-SDK/commit/8e4235abd1abe5358dea4636f89c617e2e73074a
3. Other changes are small refactors and commit a4f312908741777638663096323b5219e4bde8ea is clearing up some related dead code.

# Validation
Tested on macOS with Chrome, with a clean browser. Then toggled notification permission off and on, refreshed page and repeated a few times in different combination. Observed notification_types and token network calls were made as expected without errors.
### Checklist
   - [X] All the automated tests pass or I explained why that is not possible
   - [X] I have personally tested this on my machine or explained why that is not possible
   - [X] I have included test coverage for these changes or explained why they are not needed

**Programming Checklist**
Interfaces:
   - [X] Don't use default export
   - [X] New interfaces are in model files

Functions:
   - [X] Don't use default export
   - [X] All function signatures have return types
   - [X] Helpers should not access any data but rather be given the data to operate on.

Typescript:
   - [X] No Typescript warnings
   - [X] Avoid silencing null/undefined warnings with the exclamation point

Other:
   - [X] Iteration: refrain from using `elem of array` syntax. Prefer `forEach` or use `map`
   - [X] Avoid using global OneSignal accessor for `context` if possible. Instead, we can pass it to function/constructor so that we don't call `OneSignal.context`

### Checklist
   - [X] I have included screenshots/recordings of the intended results or explained why they are not needed

---

## Related Tickets

---

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Website-SDK/1077)

### CODE REVIEW

File: src/core/CoreModuleDirector.ts [hunk 1]
```diff
@@ -160,25 +160,39 @@ export class CoreModuleDirector {
...
+    return undefined;
+  }
+
   /**
    * Gets the current push subscription model for the current browser.
    * @returns The push subscription model for the current browser, or undefined if no push subscription exists.
    */
-  // TO DO: make synchronous by making getting the current push token synchronous
-  public async getCurrentPushSubscriptionModel(): Promise<OSModel<SupportedSubscription> | undefined> {
+  public async getPushSubscriptionModel()
```
Comment (rgomezp) [inline src/core/CoreModuleDirector.ts:189]: Did you think about ways to make getting the push token synchronous? This has the potential to make tons of dependent functions synchronous.

File: src/shared/managers/SubscriptionManager.ts [hunk 2]
```diff
@@ -163,7 +163,8 @@ export class SubscriptionManager {
   }
 
   private async _updatePushSubscriptionModelWithRawSubscription(rawPushSubscription: RawPushSubscription) {
-    const pushModel = await OneSignal.coreDirector.getCurrentPushSubscriptionModel();
+    // This undefined when it shouldn't be
```
Comment (rgomezp) [inline src/shared/managers/SubscriptionManager.ts:166]: Can you rephrase the comment?

File: src/core/CoreModuleDirector.ts [hunk 3]
```diff
@@ -160,25 +160,39 @@ export class CoreModuleDirector {
...
-
-      if (!pushToken) {
-        Log.debug("No push token found, returning undefined");
-        return undefined;
-      }
-    }
-    return this.getSubscriptionOfTypeWithToken(ModelName.PushSubscriptions, pushToken);
+    return (
+      await this.getPushSubscriptionModelByCurrentToken() ||
+      await this.getPushSubscriptionModelByLastKnownToken()
```
Comment (rgomezp) [inline src/core/CoreModuleDirector.ts:194]: I think we should add some tests for these changes

**Review by rgomezp — CHANGES_REQUESTED** [review]

File: src/core/CoreModuleDirector.ts (hunk 1 above)
Comment (jkasten2) [reply src/core/CoreModuleDirector.ts:189]: The native browse API to get the current token `registration.pushManager.getSubscription()` returns a `Promise` so it isn't possible.
* If cached it and use a cache value that could work around that, but then it won't be current and it has a risk of being out of date then.

File: src/shared/managers/SubscriptionManager.ts (hunk 2 above)
Comment (jkasten2) [reply src/shared/managers/SubscriptionManager.ts:166]: ah ops, ya that problem is solved now, as this is the reason for this PR 😄

File: src/core/CoreModuleDirector.ts (hunk 3 above)
Comment (jkasten2) [reply src/core/CoreModuleDirector.ts:194]: Added tests

**Review by emawby — APPROVED** [review]
